File: RailgunVB/Core/Events.vb
Imports AudioChord
Imports Discord
Imports Discord.WebSocket
Imports Microsoft.Extensions.DependencyInjection
Imports MongoDB.Bson
Imports RailgunVB.Core.Configuration
Imports RailgunVB.Core.Logging
Imports RailgunVB.Core.Managers
Imports RailgunVB.Core.Utilities
Imports TreeDiagram
Imports TreeDiagram.Enums
Imports TreeDiagram.Models.Server

Namespace Core
    
    Public Class Events
    
        Private ReadOnly _config As MasterConfig
        Private ReadOnly _log As Log
        Private ReadOnly _commandUtils As CommandUtils
        Private ReadOnly _serverCount As ServerCount
        Private WithEvents _client As DiscordShardedClient
        Private ReadOnly _musicService As MusicService
        Private ReadOnly _playerManager As PlayerManager
        Private ReadOnly _timerManager As TimerManager
        Private ReadOnly _services As IServiceProvider
        
        Private _initialized As Boolean = False
        Private ReadOnly _shardsReady As New Dictionary(Of Integer, Boolean)
        
        Public Sub New(services As IServiceProvider)
            _config = services.GetService(Of MasterConfig)
            _log = services.GetService(Of Log)
            _commandUtils = services.GetService(Of CommandUtils)
            _serverCount = services.GetService(Of ServerCount)
            _client = services.GetService(Of DiscordShardedClient)
            _musicService = services.GetService(Of MusicService)
            _playerManager = services.GetService(Of PlayerManager)
            _timerManager = services.GetService(Of TimerManager)
            _services = services
        End Sub
        
        Private Async Function JoinedGuildAsync(sGuild As SocketGuild) As Task Handles _client.JoinedGuild
            Await _log.LogToBotLogAsync($"<{sGuild.Name} ({sGuild.Id})> Joined", BotLogType.GuildManager)
        End Function
        
        Private Async Function LeftGuildAsync(sGuild As SocketGuild) As Task Handles _client.LeftGuild
            If _playerManager.IsCreated(sGuild.Id) Then _playerManager.GetPlayer(sGuild.Id).Player.CancelStream()
            
            Using scope As IServiceScope = _services.CreateScope()
                Dim context As TreeDiagramContext = scope.ServiceProvider.GetService(Of TreeDiagramContext)
                
                Dim sMusic As ServerMusic = Await context.ServerMusics.GetAsync(sGuild.Id)
                If sMusic IsNot Nothing AndAlso sMusic.PlaylistId <> ObjectId.Empty Then _ 
                    Await _musicService.Playlist.DeleteAsync(sMusic.PlaylistId)
                
                Await context.DeleteGuildDataAsync(sGuild.Id)
            End Using

            Await _log.LogToBotLogAsync($"<{sGuild.Name} ({sGuild.Id})> Left", BotLogType.GuildManager)
        End Function
        
        Private Async Function UserJoinedAsync(sUser As SocketGuildUser) As Task Handles _client.UserJoined
            Dim sJoinLeave As ServerJoinLeave
            
            Using scope As IServiceScope = _services.CreateScope()
                sJoinLeave = Await scope.ServiceProvider.GetService(Of TreeDiagramContext) _ 
                    .ServerJoinLeaves.GetAsync(sUser.Guild.Id)
            End Using
            
            If sJoinLeave Is Nothing Then Return
            
            Dim notification As String = sJoinLeave.GetMessage(MsgType.Join).Replace("<server>", sUser.Guild.Name) _
                .Replace("<user>", Await _commandUtils.GetUsernameOrMentionAsync(sUser))
            
            Await SendJoinLeaveMessageAsync(sJoinLeave, sUser, notification)
        End Function
        
        Private Async Function UserLeaveAsync(sUser As SocketGuildUser) As Task Handles _client.UserLeft
            Dim sJoinLeave As ServerJoinLeave
            
            Using scope As IServiceScope = _services.CreateScope()
                sJoinLeave = Await scope.ServiceProvider.GetService(Of TreeDiagramContext) _ 
                    .ServerJoinLeaves.GetAsync(sUser.Guild.Id)
            End Using
            
            If sJoinLeave Is Nothing Then Return
            
            Dim notification As String = sJoinLeave.GetMessage(MsgType.Leave)
            
            If Not (String.IsNullOrEmpty(notification)) Then notification = notification _
                .Replace("<user>", sUser.Username)
            
            Await SendJoinLeaveMessageAsync(sJoinLeave, sUser, notification)
        End Function
        
        Private Async Function SendJoinLeaveMessageAsync(data As ServerJoinLeave, user As IGuildUser, 
                                                         message As String) As Task
            If String.IsNullOrEmpty(message)
                Return
            ElseIf data.SendToDM
                Try
                    Await (Await user.GetOrCreateDMChannelAsync()).SendMessageAsync(message)
                Catch
                End Try
            ElseIf data.ChannelId <> 0
                Dim tc As ITextChannel = Await user.Guild.GetTextChannelAsync(data.ChannelId)
                Dim msg As IUserMessage = Nothing
                
                If tc IsNot Nothing
                    msg = Await tc.SendMessageAsync(message)
                End If
                
                If data.DeleteAfterMinutes > 0
                    Await Task.Run(New Action(Async Sub()
                        Await Task.Delay(TimeSpan.FromMinutes(data.DeleteAfterMinutes).TotalMilliseconds)
                        Await msg.DeleteAsync()
                    End Sub))
                End If
            End If
        End Function
        
        Private Async Function UserVoiceStateUpdatedAsync(sUser As SocketUser, before As SocketVoiceState, 
                                                          after As SocketVoiceState) As Task _
                                                          Handles _client.UserVoiceStateUpdated
            If sUser.IsBot OrElse after.VoiceChannel Is Nothing Then 
                Return
            End If
            
            Dim guild As IGuild = after.VoiceChannel.Guild
            Dim user As IGuildUser = Await guild.GetUserAsync(sUser.Id)
            
            If _playerManager.IsCreated(guild.Id) OrElse user.VoiceChannel Is Nothing Then Return
            
            Dim sMusic As ServerMusic
            Using scope As IServiceScope = _services.CreateScope()
                sMusic = Await scope.ServiceProvider.GetService(Of TreeDiagramContext) _ 
                    .ServerMusics.GetAsync(guild.Id)
            End Using
            
            If sMusic Is Nothing Then Return
            
            Dim tc As ITextChannel = If(sMusic.AutoTextChannel <> 0, Await guild.GetTextChannelAsync(sMusic.AutoTextChannel), Nothing)
            Dim vc As IVoiceChannel = user.VoiceChannel
            
            If vc.Id = sMusic.AutoVoiceChannel AndAlso tc IsNot Nothing Then _ 
                Await _playerManager.CreatePlayerAsync(user, vc, tc, True)
        End Function
        
        Private Async Function ShardReadyAsync(sClient As DiscordSocketClient) As Task Handles _client.ShardReady
            If _playerManager.PlayerContainers.Count > 0
                For Each player In _playerManager.PlayerContainers
                    player.Player.CancelStream()
                Next
            End If
            
            If Not (_shardsReady.ContainsKey(sClient.ShardId)) 
                _shardsReady.Add(sClient.ShardId, False)
            Else 
                _shardsReady(sClient.ShardId) = True
            End If
            
            Await _log.LogToConsoleAsync(new LogMessage(LogSeverity.Info, $"SHARD {sClient.ShardId}", 
                $"Shard {If(_shardsReady(sClient.ShardId), "Re-", "")}Connected! ({sClient.Guilds.Count} Servers)"))
            Await _timerManager.Initialize()
            
            If _initialized 
                Return
            ElseIf _shardsReady.Count < _client.Shards.Count
                Return
            End If
            
            _initialized = True
            _serverCount.PreviousGuildCount = _client.Guilds.Count
            
            Await _client.SetGameAsync($"{_config.DiscordConfig.Prefix}help || {_client.Guilds.Count} Servers!", 
                                       type := ActivityType.Watching)
        End Function
            
    End Class
    
End NameSpace

File: Railgun/Core/Events.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AudioChord;
using Discord;
using Discord.WebSocket;
using Microsoft.Extensions.DependencyInjection;
using MongoDB.Bson;
using Railgun.Core.Configuration;
using Railgun.Core.Logging;
using Railgun.Core.Managers;
using Railgun.Core.Utilities;
using TreeDiagram;
using TreeDiagram.Enums;
using TreeDiagram.Models.Server;
using TreeDiagram.Models.Server.Inactivity;

namespace Railgun.Core
{
	public class Events
	{
		private readonly IServiceProvider _services;
		private readonly MasterConfig _config;
		private readonly Log _log;
		private readonly DiscordShardedClient _client;
		private readonly CommandUtils _commandUtils;
		private readonly ServerCount _serverCount;
		private readonly MusicService _musicService;
		private readonly PlayerManager _playerManager;
		private readonly TimerManager _timerManager;
		private readonly InactivityManager _inactivityManager;

		private bool _initialized;
		private readonly Dictionary<int, bool> _shardsReady = new Dictionary<int, bool>();

		public Events(IServiceProvider services)
		{
			_services = services;

			_config = _services.GetService<MasterConfig>();
			_log = _services.GetService<Log>();
			_commandUtils = _services.GetService<CommandUtils>();
			_serverCount = _services.GetService<ServerCount>();
			_client = _services.GetService<DiscordShardedClient>();
			_musicService = _services.GetService<MusicService>();
			_playerManager = _services.GetService<PlayerManager>();
			_timerManager = _services.GetService<TimerManager>();
			_inactivityManager = _services.GetService<InactivityManager>();

            _client.MessageReceived += MessageReceivedAsync;
            _client.MessageUpdated += (oldMsg, newMsg, channel) => MessageReceivedAsync(newMsg);
            _client.JoinedGuild += JoinedGuildAsync;
			_client.LeftGuild += LeftGuildAsync;
			_client.UserJoined += UserJoinedAsync;
			_client.UserLeft += UserLeftAsync;
			_client.UserVoiceStateUpdated += UserVoiceStateUpdatedAsync;
			_client.ShardReady += ShardReadyAsync;
		}

		private async Task SendJoinLeaveMessageAsync(ServerJoinLeave data, IGuildUser user, string message)
		{
			if (string.IsNullOrEmpty(message)) return;
			if (data.SendToDM) {
				try {
					await (await user.GetOrCreateDMChannelAsync()).SendMessageAsync(message);
				} catch { // Ignored
				}

				return;
			}
			if (data.ChannelId == 0 && !data.SendToDM) return;

			var tc = await user.Guild.GetTextChannelAsync(data.ChannelId);

			if (tc == null) return;

			IUserMessage msg;

			try {
				msg = await tc.SendMessageAsync(message);
			} catch {
				var output = new StringBuilder()
					.AppendFormat("<{0} <{1}>> Missing Send Message Permission!", tc.Guild.Name, tc.GuildId).AppendLine()
					.AppendFormat("---- Channel Name : {0}", tc.Name).AppendLine()
					.AppendFormat("---- Channel ID   : {0}", tc.Id).AppendLine();

				await _log.LogToBotLogAsync(output.ToString(), BotLogType.TaskScheduler);
				return;
			}


			if (data.DeleteAfterMinutes > 0)
				await Task.Run(async () => {
					await Task.Delay((int)TimeSpan.FromMinutes(data.DeleteAfterMinutes).TotalMilliseconds);

					try {
						await msg.DeleteAsync();
					} catch {
						var output2 = new StringBuilder()
							.AppendFormat("<{0} <{1}>> Missing Delete Message Permission!", tc.Guild.Name, tc.GuildId).AppendLine()
							.AppendFormat("---- Channel Name : {0}", tc.Name).AppendLine()
							.AppendFormat("---- Channel ID   : {0}", tc.Id).AppendLine();

						await _log.LogToBotLogAsync(output2.ToString(), BotLogType.TaskScheduler);
					}
				});
		}

        private Task MessageReceivedAsync(SocketMessage sMessage)
        {
            if (!(sMessage is SocketUserMessage) || !(sMessage.Channel is SocketGuildChannel) || string.IsNullOrEmpty(sMessage.Content))
                return Task.CompletedTask;

            return Task.Factory.StartNew(async() => {
                var tc = (ITextChannel)sMessage.Channel;

                using (var scope = _services.CreateScope())
                {
                    var db = scope.ServiceProvider.GetService<TreeDiagramContext>();
                    var data = db.ServerInactivities.GetData(tc.GuildId);
                    var guild = tc.Guild;
                    var user = await guild.GetUserAsync(sMessage.Author.Id);

                    if (data == null) return;
	                if (!data.IsEnabled || data.InactiveDaysThreshold == 0 || data.InactiveRoleId == 0) return;
                    if (data.UserWhitelist.Any((f) => f.UserId == user.Id)) return;
                    foreach (var roleId in data.RoleWhitelist) if (user.RoleIds.Contains(roleId.RoleId)) return;

                    if (data.Users.Any((f) => f.UserId == user.Id))
                    {
	                    if (user.RoleIds.Contains(data.InactiveRoleId))
		                    await user.RemoveRoleAsync(guild.GetRole(data.InactiveRoleId));
	                    
                        data.Users.First(f => f.UserId == user.Id).LastActive = DateTime.Now;
                        return;
                    }

                    data.Users.Add(new UserActivityContainer(user.Id) { LastActive = DateTime.Now });
                }
            });
        }

		private async Task JoinedGuildAsync(SocketGuild sGuild)
			=> await _log.LogToBotLogAsync($"<{sGuild.Name} ({sGuild.Id})> Joined", BotLogType.GuildManager);

		private async Task LeftGuildAsync(SocketGuild guild)
		{
			if (_playerManager.IsCreated(guild.Id)) _playerManager.GetPlayer(guild.Id).Player.CancelStream();

			using (var scope = _services.CreateScope()) {
				var db = scope.ServiceProvider.GetService<TreeDiagramContext>();
				var data = db.ServerMusics.GetData(guild.Id);

				if (data != null && data.PlaylistId != ObjectId.Empty)
					await _musicService.Playlist.DeleteAsync(data.PlaylistId);

				db.DeleteGuildData(guild.Id);
			}

			await _log.LogToBotLogAsync($"<{guild.Name} ({guild.Id})> Left", BotLogType.GuildManager);
		}

		private async Task UserJoinedAsync(SocketGuildUser user)
		{
			ServerJoinLeave data;

			using (var scope = _services.CreateScope())
			{
				var db = scope.ServiceProvider.GetService<TreeDiagramContext>();
				var inactivityData = db.ServerInactivities.GetData(user.Guild.Id);

				if (inactivityData != null && inactivityData.IsEnabled && inactivityData.InactiveDaysThreshold != 0 && 
				    inactivityData.InactiveRoleId != 0)
				{
					if (inactivityData.Users.Any(u => u.UserId == user.Id))
						inactivityData.Users.First(u => u.UserId == user.Id).LastActive = DateTime.Now;
					else inactivityData.Users.Add(new UserActivityContainer(user.Id) { LastActive = DateTime.Now });
				}
				
				data = db.ServerJoinLeaves.GetData(user.Guild.Id);
			}

			if (data == null) return;

			var notification = data.GetMessage(MsgType.Join);

			if (string.IsNullOrEmpty(notification)) return;

			notification = notification.Replace("<server>", user.Guild.Name)
				.Replace("<user>", _commandUtils.GetUsernameOrMention(user));

			await SendJoinLeaveMessageAsync(data, user, notification);
		}

		private async Task UserLeftAsync(SocketGuildUser user)
		{
			ServerJoinLeave data;

			using (var scope = _services.CreateScope()) {
				var db = scope.ServiceProvider.GetService<TreeDiagramContext>();
				var inactivityData = db.ServerInactivities.GetData(user.Guild.Id);

				inactivityData?.Users.RemoveAll(u => u.UserId == user.Id);

				data = db.ServerJoinLeaves.GetData(user.Guild.Id);
			}

			if (data == null) return;

			var notification = data.GetMessage(MsgType.Leave);

			if (!string.IsNullOrEmpty(notification)) notification = notification.Replace("<user>", user.Username);

			await SendJoinLeaveMessageAsync(data, user, notification);
		}

		private async Task UserVoiceStateUpdatedAsync(SocketUser sUser, SocketVoiceState a1, SocketVoiceState after)
		{
			if (sUser.IsBot || after.VoiceChannel == null) return;

			var guild = (IGuild)after.VoiceChannel.Guild;
			var user = await guild.GetUserAsync(sUser.Id);

			if (_playerManager.IsCreated(guild.Id) || user.VoiceChannel == null) return;

			ServerMusic data;

			using (var scope = _services.CreateScope()) {
				data = scope.ServiceProvider.GetService<TreeDiagramContext>().ServerMusics.GetData(guild.Id);
			}

			if (data == null) return;

			var tc = data.AutoTextChannel != 0 ? await guild.GetTextChannelAsync(data.AutoTextChannel) : null;
			var vc = user.VoiceChannel;

			if (vc.Id == data.AutoVoiceChannel && tc != null) await _playerManager.CreatePlayerAsync(user, vc, tc, true);
		}

		private Task ShardReadyAsync(DiscordSocketClient sClient)
			=> Task.Factory.StartNew(async () => await ShardReadyTaskAsync(sClient));

		private async Task ShardReadyTaskAsync(DiscordSocketClient sClient) {
			if (_playerManager.PlayerContainers.Count > 0)
				foreach (var player in _playerManager.PlayerContainers) player.Player.CancelStream();

			if (!_shardsReady.ContainsKey(sClient.ShardId)) _shardsReady.Add(sClient.ShardId, false);
			else _shardsReady[sClient.ShardId] = true;

			await _log.LogToConsoleAsync(new LogMessage(LogSeverity.Info, $"SHARD {sClient.ShardId}",
				$"Shard {(_shardsReady[sClient.ShardId] ? "Re-" : "")}Connected! ({sClient.Guilds.Count} Servers)"));
			await _timerManager.InitializeAsync();
			_inactivityManager.Initialize();

			if (_initialized) return;
			if (_shardsReady.Count < _client.Shards.Count) return;

			_initialized = true;
			_serverCount.PreviousGuildCount = _client.Guilds.Count;

			await _client.SetGameAsync($"{_config.DiscordConfig.Prefix}help {Response.GetSeparator()} {_client.Guilds.Count} Servers!",
				type: ActivityType.Watching);
			await _client.SetStatusAsync(UserStatus.Online);
		}
	}
}